File: cosmos-master/code/search/interpolation_search/interpolation_search.py
def interpolationSearch(arr, n, x):
    low = 0
    high = (n - 1)

    while low <= high and x >= arr[low] and x <= arr[high]:
        mid  = low + int(((float(high - low) / ( arr[high] - arr[low])) * ( x - arr[low])))

        if arr[mid] == x:
            return mid

        if arr[mid] < x:
            low = mid + 1;

        else:
            high = mid - 1;

    return -1

def test(x):
    arr = [10, 12, 13, 16, 18, 19, 20, 21, 22, 23, 24, 33, 35, 42, 47]
    n = len(arr)

    index = interpolationSearch(arr, n, x)

    if index != -1:
        print "The element",x,"is at the index",index
    else:
        print "Element",x,"not found!"


test(123)

File: cosmos-master/code/search/interpolation_search/interpolation_search.c
#include <stdio.h>
// Part of Cosmos by OpenGenus Foundation
int interpolationSearch(int* array, int n, int key){
    int high = n-1;
    int low = 0;
    int mid;

    while ( (array[high] != array[low]) && (key >= array[low]) && (key <= array[high])) {
        mid = low + (int)( ((float)(high-low) / (array[high] - array[low])) * (key - array[low]) );

        if(array[mid] == key)
            return mid;

        else if (array[mid] < key)
            low = mid + 1;

        else
            high = mid - 1;

    }
    return -1;
}

void test(){

    // Array of items on which search will be conducted.
    int array[10] = {1, 5, 9, 12, 16, 18, 25, 56, 76, 78};
    int n = sizeof(array)/sizeof(array[0]);

    // Element to be searched
    int key = 25;

    int index = interpolationSearch(array, n, key);

    if(index !=-1)
        printf("Element found at index %d\n", index);
    else
        printf("EElement not found.\n");
}

int main() {
    test(25);
    return 0;
}
